lib/libvfio-user.c

```diff
@@ -920,133 +894,284 @@ static int
-        vfu_log(vfu_ctx, LOG_ERR, "migration not configured");
-        return ERROR_INT(ENOTSUP);
+    struct vfio_user_device_feature *req = msg->in.iov.iov_base;
+
+    uint32_t operations = req->flags & ~VFIO_DEVICE_FEATURE_MASK;
+    uint32_t feature = req->flags & VFIO_DEVICE_FEATURE_MASK;
+
+    uint32_t supported_ops = device_feature_flags_supported(vfu_ctx, feature);
+
+    if ((req->flags & supported_ops) != operations || supported_ops == 0) {
+        return ERROR_INT(EINVAL);
     }
 
-    switch (dirty_pages->flags) {
-    case VFIO_IOMMU_DIRTY_PAGES_FLAG_START:
-        ret = dma_controller_dirty_page_logging_start(vfu_ctx->dma,
-                  migration_get_pgsize(vfu_ctx->migration));
-        break;
+    ssize_t ret;
+
+    if (req->flags & VFIO_DEVICE_FEATURE_PROBE) {
+        msg->out.iov.iov_len = msg->in.iov.iov_len;
+
+        if (req->argsz < msg->out.iov.iov_len) {
+            msg->out.iov.iov_len = 0;
+            return ERROR_INT(EINVAL);
+        }
+
+        msg->out.iov.iov_base = malloc(msg->out.iov.iov_len);
+
+        if (msg->out.iov.iov_base == NULL) {
+            return ERROR_INT(ENOMEM);
+        }
+
+        memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base,
+               msg->out.iov.iov_len);
 
-    case VFIO_IOMMU_DIRTY_PAGES_FLAG_STOP:
-        dma_controller_dirty_page_logging_stop(vfu_ctx->dma);
         ret = 0;
-        break;
+    } else if (req->flags & VFIO_DEVICE_FEATURE_GET) {
+        if (is_migration_feature(feature)) {
+            ret = handle_migration_device_feature_get(vfu_ctx, msg, req);
+        } else if (is_dma_feature(feature)) {
+            ret = handle_dma_device_feature_get(vfu_ctx, msg, req);
+        } else {
+            return ERROR_INT(EINVAL);
+        }
+    } else if (req->flags & VFIO_DEVICE_FEATURE_SET) {
+        msg->out.iov.iov_len = msg->in.iov.iov_len;
 
-    case VFIO_IOMMU_DIRTY_PAGES_FLAG_GET_BITMAP:
-        ret = handle_dirty_pages_get(vfu_ctx, msg);
-        break;
+        if (req->argsz < msg->out.iov.iov_len) {
+            msg->out.iov.iov_len = 0;
+            return ERROR_INT(EINVAL);
+        }
 
-    default:
-        vfu_log(vfu_ctx, LOG_ERR, "bad flags %#x", dirty_pages->flags);
-        ret = ERROR_INT(EINVAL);
-        break;
+        msg->out.iov.iov_base = malloc(msg->out.iov.iov_len);
+
+        if (msg->out.iov.iov_base == NULL) {
+            return ERROR_INT(ENOMEM);
+        }
+
+        memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base,
+               msg->out.iov.iov_len);
+
+        struct vfio_user_device_feature *res = msg->out.iov.iov_base;
+
+        if (is_migration_feature(feature)) {
+            ret = handle_migration_device_feature_set(vfu_ctx, feature, res);
+        } else if (is_dma_feature(feature)) {
+            ret = handle_dma_device_feature_set(vfu_ctx, feature, res);
+        } else {
+            return ERROR_INT(EINVAL);
+        }
+    } else {
+        return ERROR_INT(EINVAL);
```

Comment:
log debug message

Reply:
I've refactored so this case isn't there anymore.